### PR TITLE
Redesign query sanitizer (#618)

### PR DESCRIPTION
**Pull request by Mbucari**
It's often the most piddly shit that take the most time and effort to solve, and #618 is a prime example.

First of all, I was wrong about QueryParser converting everything to a number. That was done by `SearchEngine.FormatSearchQuery` The regex was being too aggressive, and the only way to _not_ convert ASINs to integers was to tokenize the whole query and not do it for an ID. Thank god Lucene exposes their built-in tokenizer.

But since I was tokenizing the whole query and looking at each term anyway, I took the opportunity to dispense with a lot of the regex that could be done more easily in the token loop.

### CODE REVIEW

**Comment by Mbucari** [comment]
I see a nice way to simplify the "[tag]" expanded. Converting to draft.

**Comment by Mbucari** [comment]
Ready for review!

**Comment by rmcrackan** [comment]
> It's often the most piddly shit that take the most time and effort to solve

Truer words. Add this new huge feature? It's actually a minor refactor and new implementation of an existing well implemented feature -- done. Change the text on this button? Oh fuck, it has to go through the localization engine which has a hack for if you're pluralizing words with a capital letter in the middle and you can't use that exact string because it contains a word reserved by our ORM extension methods...

I haven't touched this part of the search engine in ages; I'm sure you can tell what a colossal effort it was at the time. I'm so glad I made those unit tests. They were indispensable while I was learning the weird nuances of lucene and they also let us run regressions during refactors like this.

Can you throw in some unit tests for your asin fixes?

**Comment by rmcrackan** [comment]
RE unit tests, I'm thinking of: case-specific, leading numerals, and numerals-only IDs

**Comment by rmcrackan** [comment]
> I took the opportunity to dispense with a lot of the regex

They were a carefully crafted solution that served me well for years, but I'm *never* surprised when regex is replaced with a specialized parser

**Comment by rmcrackan** [comment]
Unrelated: WheelComboBox.axaml.cs has some notifications at the bottom of the PR. Looks like notes from the new version of avalonia

**Comment by Mbucari** [comment]
> I'm sure you can tell what a colossal effort it was at the time.

I can. It was so hard to follow and I could feel your frustration reaching out across time and space.

> I'm so glad I made those unit tests.

I'm glad you made them to. They made my job easier, but it was still a royal PITA

> RE unit tests

Done.

> WheelComboBox.axaml.cs

Done.
